### PR TITLE
Decode generated token_ids incrementally

### PR DESCRIPTION
**Pull request by AllentDan**
Tested OK with InternLM, Llama2, Qwen, Puyu, Vicuna

### CODE REVIEW

File: lmdeploy/turbomind/tokenizer.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ def encode(self, s: str):
             add_eos = True
         return self.model.Encode(s, add_bos=add_bos, add_eos=add_eos)
 
-    def decode(self, t: Sequence[int]):
+    def decode(self, t: Sequence[int], offset: Optional[int] = None):
```
Comment (lvhan028) [inline lmdeploy/turbomind/tokenizer.py:53]: Please update docstring as well

**Comment by lvhan028** [comment]
May test puyu 100B. It uses SentencePieceTokenizer

File: lmdeploy/turbomind/tokenizer.py [hunk 2]
```diff
@@ -139,16 +159,23 @@ def encode(self, s: str):
...
         Args:
             t (List[int]): a list of token ids
+            offset (int): for incrementally decoding. Default to None, which
+                means not applied.
         Returns:
             str: text of decoding tokens
         """
         skip_special_tokens = True
-        return self.model.decode(t, skip_special_tokens=skip_special_tokens)
+        t = t[offset:] if offset else t
```
Comment (grimoire) [inline lmdeploy/turbomind/tokenizer.py:173]: `t[None:] == t`. This branch can be removed.

File: lmdeploy/turbomind/tokenizer.py [hunk 3]
```diff
@@ -139,16 +186,23 @@ def encode(self, s: str):
...
         Args:
             t (List[int]): a list of token ids
+            offset (int): for incrementally decoding. Default to None, which
+                means not applied.
         Returns:
             str: text of decoding tokens
         """
         skip_special_tokens = True
-        return self.model.decode(t, skip_special_tokens=skip_special_tokens)
+        t = t[offset:]
```
Comment (wangruohui) [inline lmdeploy/turbomind/tokenizer.py:200]: same as above.

**Review by wangruohui — APPROVED** [review]

File: lmdeploy/turbomind/tokenizer.py [hunk 4]
```diff
@@ -95,6 +120,8 @@ def __init__(self, model_dir: str):
                   'It may take long time to initialize the tokenizer.')
         self.model = AutoTokenizer.from_pretrained(model_dir,
                                                    trust_remote_code=True)
+        self.need_padding = isinstance(self.model, LlamaTokenizerFast)
```
Comment (wangruohui) [inline lmdeploy/turbomind/tokenizer.py:123]: Do we only support llama in huggingface model? Or is it better to have a list of "need_padding" tokenizers?

Comment (AllentDan) [reply lmdeploy/turbomind/tokenizer.py:123]: Other tokenizer can automatically pad white space.

File: lmdeploy/turbomind/tokenizer.py [hunk 5]
```diff
@@ -50,17 +69,23 @@ def encode(self, s: str):
...
             t (List[int]): a list of token ids
+            offset (int): for incrementally decoding. Default to None, which
+                means not applied.
         Returns:
             str: text of decoding tokens
         """
         if isinstance(t, torch.Tensor):
             t = t.tolist()
-        return self.model.Decode(t)
+        t = t[offset:]
```
[the remark this answers is not in the document]
Comment (AllentDan) [reply lmdeploy/turbomind/tokenizer.py:84]: empty is valid here and wil return ''

File: lmdeploy/serve/turbomind/chatbot.py [hunk 6]
```diff
@@ -629,30 +627,29 @@ def stream_consumer(postprocess, res_queue, session, n_input_token,
...
-                                         sequence_length)
+                output_str = postprocess(
+                    output_ids, np.array([[n_token]], dtype=np.uint32))
+                n_token = output_ids.shape[-1]
                 text = output_str[0].decode()
                 if display:
-                    new_text = text[len(session.response):]
-                    print(new_text, end='', flush=True)
+                    print(text, end='', flush=True)
                 session.response = text
```
Comment (lvhan028) [inline lmdeploy/serve/turbomind/chatbot.py:650]: Chatbot should return all of the generated text
```
session.response = session.response + text
```

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**